### PR TITLE
Feat issue 17 logistic selectivity module

### PR DESCRIPTION
**Pull request by Andrea-Havron-NOAA**
**What is the feature?** 
The logistic selectivity module

**How have you implemented the solution?** 
- new [fims_math.hpp](https://github.com/NOAA-FIMS/FIMS/blob/feat-issue-17-logistic-selectivity-module/inst/include/fims_math.hpp) file under [inst/include](https://github.com/NOAA-FIMS/FIMS/tree/feat-issue-17-logistic-selectivity-module/inst/include)
  - logistic() function added as an inline const T
- new [functors](https://github.com/NOAA-FIMS/FIMS/tree/feat-issue-17-logistic-selectivity-module/inst/include/population_dynamics/selectivity/functors) folder under [inst/include/population_dynamics/selectivity](https://github.com/NOAA-FIMS/FIMS/tree/feat-issue-17-logistic-selectivity-module/inst/include/population_dynamics/selectivity)
  - [selectivity_base.hpp](https://github.com/NOAA-FIMS/FIMS/blob/feat-issue-17-logistic-selectivity-module/inst/include/population_dynamics/selectivity/functors/selectivity_base.hpp): develop initial selectivity base struct with evaluate function
  - [logistic.hpp](https://github.com/NOAA-FIMS/FIMS/blob/feat-issue-17-logistic-selectivity-module/inst/include/population_dynamics/selectivity/functors/logistic.hpp): develop logistic selectivity struct that assigns the logistic function from fims_math to the evaluate function call
- edited [inst/include/population_dynamics/selectivity.hpp](https://github.com/NOAA-FIMS/FIMS/blob/feat-issue-17-logistic-selectivity-module/inst/include/population_dynamics/selectivity/selectivity.hpp): added `#include "functors/selectivity_base.hpp"` to link changes

**Does it impact any other area of the project? Please describe which features are impacted by this change.**
This change does not impact any other area of the project.

**Does this change impact the model input or output? Please describe any way this change impacts the model input or output.**
This change does not impact the model input or output as the selectivity module is not yet linked up to the objective function, [src/FIMS.cpp](https://github.com/NOAA-FIMS/FIMS/blob/feat-issue-17-logistic-selectivity-module/src/FIMS.cpp)

**How to test this change Please include a test file and/or Github workflow.**
- [tests/gtest/test/test_fims_math.cpp](https://github.com/NOAA-FIMS/FIMS/blob/feat-issue-17-logistic-selectivity-module/tests/gtest/test_fims_math.cpp): tests added to test the logistic function in fims_math.hpp

### CODE REVIEW

**Review by ChristineStawitz-NOAA — CHANGES_REQUESTED** [review]
test

File: inst/include/fims_math.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#ifndef FIMS_MATH_HPP
+#define FIMS_MATH_HPP
+
+//note: this is modeling platform specific, must be controlled by
+//preprocessing macros
+#include "def.hpp"
+#include <cmath>
+
+namespace fims {
+
```
Comment (JaneSullivan-NOAA) [inline inst/include/fims_math.hpp:40]: minimize use of whitespace https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace

File: inst/include/population_dynamics/selectivity/functors/logistic.hpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+#include "../../../interface/interface.hpp"
+#include "../../../fims_math.hpp"
+#include "selectivity_base.hpp"
+
+namespace fims {
+
+template<typename T>
+struct LogisticSelectivity : public SelectivityBase<T> {
+    T a50;
```
Comment (JaneSullivan-NOAA) [inline inst/include/population_dynamics/selectivity/functors/logistic.hpp:43]: Has the group decided on 'T' vs 'Type'?

File: inst/include/population_dynamics/selectivity/functors/logistic.hpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    T slope;
+
+    LogisticSelectivity(): SelectivityBase<T>(){
+
+    }
+
+    virtual const T evaluate(const T& age) {
+        return fims::logistic<T>(a50, slope, age);
+    }
+};
```
Comment (JaneSullivan-NOAA) [inline inst/include/population_dynamics/selectivity/functors/logistic.hpp:53]: is this semicolon needed?

File: inst/include/fims_math.hpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+#ifdef USE_STANDARD_LIB
+
+
+
```
Comment (EJDick-NOAA) [inline inst/include/fims_math.hpp:45]: Needs documentation

File: inst/include/fims_math.hpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    return std::exp(x);
+}
+
+template <class T>
+inline const T log(const T& x) {
+    return std::log(x);
+}
+
+template <class T>
+inline const T logistic(const T& mean, const T& slope, const T& x) {
```
Comment (Bai-Li-NOAA) [inline inst/include/fims_math.hpp:57]: Define terms used in code line 58. In particular, mean and slope. “Mean” is an index of 50% of function output, probably better termed median. “Slope” is typically a multiplicative scalar and not an inverse multiplier, probably better termed “inverse slope”

File: inst/include/fims_math.hpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    return (1.0) / (1.0 + exp(-1.0 * (x - mean) / slope));
```
Comment (Bai-Li-NOAA) [inline inst/include/fims_math.hpp:58]: The logistic equation does not match the FIMS design specification document (https://noaa-fims.github.io/collaborative_workflow/m1-model-specification.html#logistic-function-with-extensions); need to use difference in the index at 50% and the index at 95% for the parameterization as in the workflow specs

File: inst/include/population_dynamics/selectivity/functors/logistic.hpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+struct LogisticSelectivity : public SelectivityBase<T> {
+    T a50;
+    T slope;
+
+    LogisticSelectivity(): SelectivityBase<T>(){
+
+    }
+
+    virtual const T evaluate(const T& age) {
+        return fims::logistic<T>(a50, slope, age);
```
Comment (EJDick-NOAA) [inline inst/include/population_dynamics/selectivity/functors/logistic.hpp:51]: Consider having users input a50 & a95, with estimation using slope parameterization

**Review by EJDick-NOAA — CHANGES_REQUESTED** [review]
1. Fix issue with the build; test case number 2 exp.use_integer_inputs failed

- [ ] The code is well-designed.
- [ ] The functionality is good for the users of the code.
- [ ] Any User Interface changes are sensible and look good.
- [ ] Any parallel programming is done safely.
- [ ] The code isn’t more complex than it needs to be.
- [ ] The developer isn’t implementing things they might need in the future but don’t know they need now.
- [ ] Code has appropriate unit tests.
- [ ] Tests are well-designed.
- [ ] The developer used clear names for everything.
- [ ] Comments are clear and useful, and mostly explain why instead of what.
- [ ] Code is appropriately documented (doxygen and roxygen).
- [ ] The code conforms to our style guides.

File: tests/gtest/test_fims_math.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  {
+    EXPECT_NEAR(fims::log(3.0), 1.098612, 0.0001);
+    EXPECT_EQ(fims::log(2.5), std::log(2.5));
+    EXPECT_TRUE(std::isnan(fims::log(-2.5)));
+  }
+  
+  // Test logistic with double values  
+  TEST(logistic, function)
+  {
+    double mean = 3.0;
```
Comment (Bai-Li-NOAA) [inline tests/gtest/test_fims_math.cpp:38]: Expand the test suite and try different input values.

**Comment by Bai-Li-NOAA** [comment]
> Test number 2 is failing in build workflow

I updated the test and created an issue #41 for resolving the failed test.

**Review by ChristineStawitz-NOAA — APPROVED** [review]
approving Bai's testing changes!

**Review by iantaylor-NOAA — COMMENTED** [review]
Matching the parameterization used in the model comparison project seems reasonable but differs from our model specification document: https://noaa-fims.github.io/collaborative_workflow/m1-model-specification.html#logistic-function-with-extensions. 

I'm a fan of the parameterization in the collaborative workflow doc, and I think that either one  will work fine for replicating the model comparison project results, but I also see the simplicity of comparing parameters within the same parameterization which this commit enables.

[edit: to clarify, it's recent commit 7ce1694c00056713791a1ee2d6d9b62c104652d0 that I'm referring to--probably should have commented on the individual lines instead of as a code review comment]

**Pull request closed without merging**